### PR TITLE
Use the name of the table to create the relation

### PR DESCRIPTION
**Pull request by thiagotalma**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌

Use the real name of the table to create the relationship when there is a composite primary key.

Prevents the generation of random relationship names.

**Turn this:**

```PHP
<?php

(...)

/**
 *
 * @property int $id
 * @property int $employee_id
 *
 * @property Employee $employee
 * @property Route $employee0 <---
 */
class Company extends \yii\db\ActiveRecord
{
    (...)

    /**
     * Gets query for [[Employee]].
     *
     * @return \yii\db\ActiveQuery
     */
    public function getEmployee()
    {
        return $this->hasOne(Employee::class, ['id' => 'employee_id']);
    }

    /**
     * Gets query for [[Employee0]]. <---
     *
     * @return \yii\db\ActiveQuery
     */
    public function getEmployee0() /* <--- */
    {
        return $this->hasOne(Route::class, ['employee_id' => 'employee_id', 'id' => 'route_id']);
    }
}
```

**Into this:**

```PHP
<?php

(...)

/**
 *
 * @property int $id
 * @property int $employee_id
 * @property string $route_id
 *
 * @property Employee $employee
 * @property Route $route <---
 */
class Company extends \yii\db\ActiveRecord
{
    (...)

    /**
     * Gets query for [[Employee]].
     *
     * @return \yii\db\ActiveQuery
     */
    public function getEmployee()
    {
        return $this->hasOne(Employee::class, ['id' => 'employee_id']);
    }

    /**
     * Gets query for [[Route]]. <---
     *
     * @return \yii\db\ActiveQuery
     */
    public function getRoute() /* <--- */
    {
        return $this->hasOne(Route::class, ['employee_id' => 'employee_id', 'id' => 'route_id']);
    }
}
```

### CODE REVIEW

**Comment by samdark** [comment]
What if both relations are needed?

**Comment by thiagotalma** [comment]
Look, I'm not eliminating the relationship.

Just renaming. Using the table name instead of the primary key.

**Comment by samdark** [comment]
Tests should be fixed.

**Comment by thiagotalma** [comment]
> Tests should be fixed.

@samdark Done!

**Review by xepozz — APPROVED** [review]

**Comment by thiagotalma** [comment]
@samdark Is there any chance this change will be present in the next release?

**Comment by xepozz** [comment]
@bizley @terabytesoftw Could you please review?

**Review by terabytesoftw — APPROVED** [review]

**Comment by rhertogh** [comment]
I think there should be a configuration option for this change for backwards compatibility. E.g. when used with https://github.com/schmunk42/yii2-giiant.

**Review by bizley — APPROVED** [review]

**Comment by bizley** [comment]
I would not want to keep the old way. @schmunk42 could you check if this change is problematic with your package?

**Comment by schmunk42** [comment]
One thing is, that the workflow with `giiant` is to regenerate base-models everytime your db changes, which would be problematic here, since the relations would change.

But another more important thing is, that if you have a table `company` with 2 FKs to Employee eg. `president_id` and `vice_president_id`, and you take the relation name from the table name you'll end up with `getEmployee` and `getEmployee0`. In this use-case you are better off with relation names from the column.

I'd not change this. At least not for yii2.
Or make it configurable.

**Comment by bizley** [comment]
Ok, I see. If we correct this in one case, the other one is broken. @thiagotalma could you make it configurable?

**Comment by thiagotalma** [comment]
Configuration in the module or on the gii screen?

**Comment by schmunk42** [comment]
gii screen I think

File: src/generators/model/Generator.php [hunk 1]
```diff
@@ -153,6 +154,7 @@ public function attributeLabels()
             'generateRelations' => 'Generate Relations',
             'generateJunctionRelationMode' => 'Generate Junction Relations As',
             'generateRelationsFromCurrentSchema' => 'Generate Relations from Current Schema',
+            'generateRelationNameFromDestinationTable' => 'Generate Relations Name Using Destination Table Name',
```
Comment (thiagotalma) [inline src/generators/model/Generator.php:157]: I'll need help with the variable name and the hint.

I am not fluent in English.

Comment (bizley) [reply src/generators/model/Generator.php:157]: Maybe `Generate Relation Names Using Target Table Name`

**Comment by schmunk42** [comment]
@thiagotalma There should be tests for both cases (original ones and the new one).

**Comment by thiagotalma** [comment]
Can anyone help me with the tests, please?
(Yes, it's a shame. I know)

**Comment by schmunk42** [comment]
You might need an additional param in your data provider:

```
public function testRelations($tableName, $fileName, $useClassConstant, $relations, $fromDestTable = false)
    {
        $generator = new ModelGenerator();
...
        $generator->generateRelationNameFromDestinationTable = $fromDestTable;
```

And use it in your new use-cases in `relationsProvider()`:
```
            ['product_language', 'ProductLanguage.php', false, [
                [
                    'name' => 'function getSupplier()',
                    'name' => 'function getProduct()',
                    'relation' => "\$this->hasOne(Product::className(), ['supplier_id' => 'supplier_id', 'id' => 'id']);",
                    'expected' => true,
                ],
                [
                    'name' => 'function getSupplier0()',
                    'name' => 'function getSupplier()',
                    'relation' => "\$this->hasOne(Supplier::className(), ['id' => 'supplier_id']);",
                    'expected' => true,
                ],
            ],
         true // $fromDestTable
         ],
``` 

The existing ones should be kept/reverted.

**Comment by rhertogh** [comment]
> But another more important thing is, that if you have a table `company` with 2 FKs to Employee eg. `president_id` and `vice_president_id`, and you take the relation name from the table name you'll end up with `getEmployee` and `getEmployee0`. In this use-case you are better off with relation names from the column.
> 
> I'd not change this. At least not for yii2. Or make it configurable.

I was thinking about this some more since I encountered a similar situation today.
In my opinion the best option would be to use the name of the foreign key itself.
In the example from @schmunk42 above the foreign keys probably would be named "president" and "vice_president" respectively.

So my proposal would be to make it indeed configurable as discussed above and add a third option.
The configuration would be `$generateRelationNameFrom`, with the options:
- column name
- target table name
- foreign key name

**Comment by thiagotalma** [comment]
Done!

Thanks, @schmunk42

**Comment by schmunk42** [comment]
@thiagotalma You're welcome.

LGTM now :)

**Review by bizley — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

**Comment by bizley** [comment]
Oh, indeed. @thiagotalma could you prepare another PR correcting this?

**Comment by thiagotalma** [comment]
@bizley 
Done!
https://github.com/yiisoft/yii2-gii/pull/544

**Comment by thiagotalma** [comment]
Hello guys!

What do you think about releasing the new version?
The milestone is already cleared.

I'm just waiting for these changes to move forward with my project.

@samdark

**Pull request merged**